### PR TITLE
[helpers] Better handling of  different sequence dimensions

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Related to #138
- [x] Check in the block unit test that we handle different query and key length in the encoder/decoder case
- [x] Fix the rotary embedding handling of different query/key lengths
- [x] Expose a property in attentions to state whether different query/key lengths are supported
- [x] Improve the error messages, catch properly unsupported cases

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -152,6 +152,7 @@ Patrick et al., 2021](https://arxiv.org/abs/2106.05392)*
 
 - [Sine](xformers/components/positional_embedding/sine.py)
 - [Vocabulary](xformers/components/positional_embedding/vocab.py)
+- [Rotary](xformers/components/positional_embedding/rotary.py)
```
Comment (blefaudeux) [inline README.md:155]: unrelated, I just forgot that back when rotary was introduced

File: tests/test_attentions.py [hunk 2]
```diff
@@ -237,17 +237,12 @@ def test_different_kq_dimensions(
...
-        "local",
-        "random",
-        "lambda",
-        "linformer",
-        "blocksparse",
-    }:
+
+    multi_head = _get_multihead(attention_name, 0.0, 0.0, False, heads, device)
+
+    if multi_head.attention.requires_same_k_q_dimensions:
```
Comment (blefaudeux) [inline tests/test_attentions.py:243]: a bit cleaner I think, add an attribute to attentions to report whether or not they support having different query and key lengths

File: tests/test_block_factory.py [hunk 3]
```diff
@@ -211,3 +211,16 @@ def test_xformer_decoder_block(
 
     encoded = encoder_block(inputs)
     _ = decoder_block(inputs, encoded, encoder_att_mask=att_mask, input_mask=input_mask)
+
+    # Test different sequence lengths when encoding and decoding
```
Comment (blefaudeux) [inline tests/test_block_factory.py:215]: check that either we support or assert of all the possible combinations

File: xformers/components/multi_head_dispatch.py [hunk 4]
```diff
@@ -146,6 +146,18 @@ def forward(
         B, S_Q, _ = query.size()  # Batch x Sequence x Embedding (latent)
         _, S_K, _ = key.size()  # K, Q's sequence length could differ
 
+        # Catch different query and key length but a causal attention
+        if S_Q != S_K:
```
Comment (blefaudeux) [inline xformers/components/multi_head_dispatch.py:150]: Makes it a little more understandable I think, catch early a combination which is not supported

File: xformers/components/positional_embedding/rotary.py [hunk 5]
```diff
@@ -73,7 +78,10 @@ def forward(
         self, q: torch.Tensor, k: torch.Tensor
     ) -> Tuple[torch.Tensor, torch.Tensor]:
         self._cos_cached, self._sin_cached = self._update_cos_sin_tables(
-            q, seq_dimension=-2
+            k, seq_dimension=-2
         )
 
-        return apply_rotary_pos_emb(q, k, self._cos_cached, self._sin_cached)
+        return (
```
Comment (blefaudeux) [inline xformers/components/positional_embedding/rotary.py:84]: side effect, out of principle rotary supports different k/q dimensions but the previous implementation would choke on it, fixing that

**Review by dianaml0 — APPROVED** [review]
Really nice to have all the assertions and testing, LGTM!

**Pull request merged**